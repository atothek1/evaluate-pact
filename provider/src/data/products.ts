const faker = require("faker")
import { Product } from "../../src/types";

const { commerce, datatype, image } = faker;
faker.seed(5678);

function createImageUrls(): string[] {
    const res: string[] = [];
    for (let i = 0; i < 10; i++) {
        res.push(image.imageUrl());
    }
    return res
}

export function createProduct(): Product {
    return {
        id: datatype.uuid(),
        name: commerce.productName(),
        price: parseFloat(commerce.price()) * 10,
        description: commerce.productDescription(),
        isAvailable: datatype.boolean(),
        imageUrls: createImageUrls()
    }
}
export class Products {
    constructor(private products: Product[] = []) { }
    clear() { this.products = [] }
    add(...products: Product[]) { this.products.push(...products) }
    get() { return [ ...this.products ] }
}
export const productRepository = new Products();