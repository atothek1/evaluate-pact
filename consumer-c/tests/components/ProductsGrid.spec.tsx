import React from "react";
import { ProductsGrid } from "../../src/components/products/ProductsGrid";
import { render, waitFor } from "../utils";
import { pactWith } from 'jest-pact';
import { filteredProductsRequest, productsRequest, withListOfProductsResponse, withOneAvailableProductInTheListOfProductsResponse } from "./utils/productRequests";

pactWith({ consumer: 'e2e Consumer C', provider: 'e2e Provider', cors: true }, provider => {

    describe("getProducts()", () => {
        beforeEach(async () => {
            await provider.removeInteractions()
            process.env.API_HOST = provider.mockService.baseUrl;
        });
        
        test("integration without filters", async () => {
            await provider.addInteraction({
                ...productsRequest,
                ...withListOfProductsResponse
            })
            const { baseElement, getByRole } = render(<ProductsGrid />);
            expect(getByRole("heading")).toHaveTextContent("FETCHING");

            await waitFor(() => {
                getByRole("list")
            });

            expect(getByRole("list").children).toHaveLength(10);
            // not very usefull i know just for fiddling around
            expect(baseElement).toMatchSnapshot();
        })

        test("with filters", async () => {
            await provider.addInteraction({
                ...filteredProductsRequest({ isAvailable: "true" }),
                ...withOneAvailableProductInTheListOfProductsResponse()
            })

            const { baseElement, getByRole } = render(<ProductsGrid isAvailable={true} />);
            expect(getByRole("heading")).toHaveTextContent("FETCHING");

            await waitFor(() => {
                getByRole("list")
            });

            expect(getByRole("list").children).toHaveLength(1);
            // not very usefull i know just for fiddling around
            expect(baseElement).toMatchSnapshot();
        })
    })
})