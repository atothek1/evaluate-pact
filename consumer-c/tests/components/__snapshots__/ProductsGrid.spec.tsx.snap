// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pact between e2e Consumer C and e2e Provider with 30000 ms timeout for Pact getProducts() integration without filters 1`] = `
<body>
  <div>
    <h2>
      Products
    </h2>
    <ul>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Sleek Soft Chips
          </h2>
          <p>
            Boston's most advanced compression wear technology increases muscle oxygenation, stabilizes active muscles
          </p>
          <p
            style="font-weight: bold;"
          >
            41.7
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Licensed Plastic Tuna
          </h2>
          <p>
            The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality
          </p>
          <p
            style="font-weight: bold;"
          >
            84.4
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Tasty Cotton Pants
          </h2>
          <p>
            Boston's most advanced compression wear technology increases muscle oxygenation, stabilizes active muscles
          </p>
          <p
            style="font-weight: bold;"
          >
            99.3
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Generic Frozen Gloves
          </h2>
          <p>
            The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality
          </p>
          <p
            style="font-weight: bold;"
          >
            57.7
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Small Metal Mouse
          </h2>
          <p>
            New range of formal shirts are designed keeping you in mind. With fits and styling that will make you stand apart
          </p>
          <p
            style="font-weight: bold;"
          >
            95.1
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Intelligent Wooden Soap
          </h2>
          <p>
            Ergonomic executive chair upholstered in bonded black leather and PVC padded seat and back for all-day comfort and support
          </p>
          <p
            style="font-weight: bold;"
          >
            25.7
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Awesome Frozen Chair
          </h2>
          <p>
            The Apollotech B340 is an affordable wireless mouse with reliable connectivity, 12 months battery life and modern design
          </p>
          <p
            style="font-weight: bold;"
          >
            4
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Awesome Granite Gloves
          </h2>
          <p>
            The automobile layout consists of a front-engine design, with transaxle-type transmissions mounted at the rear of the engine and four wheel drive
          </p>
          <p
            style="font-weight: bold;"
          >
            72.6
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Small Fresh Pizza
          </h2>
          <p>
            The beautiful range of Apple Naturalé that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients
          </p>
          <p
            style="font-weight: bold;"
          >
            29.3
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Incredible Soft Bacon
          </h2>
          <p>
            The Apollotech B340 is an affordable wireless mouse with reliable connectivity, 12 months battery life and modern design
          </p>
          <p
            style="font-weight: bold;"
          >
            71.2
          </p>
        </div>
      </li>
    </ul>
  </div>
</body>
`;

exports[`Pact between e2e Consumer C and e2e Provider with 30000 ms timeout for Pact getProducts() with filters 1`] = `
<body>
  <div>
    <h2>
      Products
    </h2>
    <ul>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Fantastic Concrete Towels
          </h2>
          <p>
            The beautiful range of Apple Naturalé that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients
          </p>
          <p
            style="font-weight: bold;"
          >
            90.5
          </p>
        </div>
      </li>
    </ul>
  </div>
</body>
`;
